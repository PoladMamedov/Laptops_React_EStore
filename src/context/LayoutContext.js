import React, { useContext, useState } from "react";
const LayoutContext = React.createContext();
const LayoutContextUpdate = React.createContext();

export function useLayout() {
  return useContext(LayoutContext);
}
export function useLayoutUpdate() {
  return useContext(LayoutContextUpdate);
}

export function LayoutProvider({ children }) {
  const [isListLayout, setIsListLayout] = useState(false);

  function setCardLayout() {
    setIsListLayout(false);
  }
  function setListLayout() {
    setIsListLayout(true);
  }

  return (
    <LayoutContext.Provider value={isListLayout}>
      <LayoutContextUpdate.Provider value={{ setListLayout, setCardLayout }}>{children}</LayoutContextUpdate.Provider>
    </LayoutContext.Provider>
  );
}
