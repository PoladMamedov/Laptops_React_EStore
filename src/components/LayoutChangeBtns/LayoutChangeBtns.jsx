import { useEffect } from "react";
import { useLayout, useLayoutUpdate } from "../../context/LayoutContext";
import { useWindowWidth } from "../../hooks/useWindowWidth";

export function LayoutChangeBtns() {
  const isListlayout = useLayout();
  const { setListLayout, setCardLayout } = useLayoutUpdate();
  const windowWidth = useWindowWidth();
  useEffect(() => {
    if (localStorage.getItem("isListLayout") === "false") {
      localStorage.setItem("islistLayout", "false");
      setCardLayout();
    }
    if (localStorage.getItem("isListLayout") === "true") {
      localStorage.setItem("islistLayout", "true");
      setListLayout();
    }
    if (windowWidth < 580) setCardLayout();
  });
  return windowWidth < 580 ? null : (
    <div data-testid="layout-btns" className="btn-group">
      <input type="checkbox" className="btn-check" checked={isListlayout ? false : true} readOnly />
      <label
        onClick={() => {
          localStorage.setItem("isListLayout", "false");
          setCardLayout();
        }}
        className="btn btn-outline-secondary"
      >
        Cards
      </label>
      <input type="checkbox" className="btn-check" checked={isListlayout ? true : false} readOnly />
      <label
        onClick={() => {
          localStorage.setItem("isListLayout", "true");
          setListLayout();
        }}
        className="btn btn-outline-secondary"
      >
        List
      </label>
    </div>
  );
}
