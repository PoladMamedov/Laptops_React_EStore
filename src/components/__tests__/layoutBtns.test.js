import "@testing-library/jest-dom";
import { render, screen, cleanup } from "@testing-library/react";
import { LayoutChangeBtns } from "../LayoutChangeBtns/LayoutChangeBtns";
import { LayoutProvider } from "../../context/LayoutContext";
import renderer from "react-test-renderer";

afterEach(cleanup);

test("should render proper layout buttons", () => {
  render(
    <LayoutProvider>
      <LayoutChangeBtns />
    </LayoutProvider>
  );

  expect(screen.getByTestId("layout-btns")).toBeInTheDocument();
  expect(screen.getByTestId("layout-btns")).toHaveTextContent("Card");
  expect(screen.getByTestId("layout-btns")).toHaveTextContent("List");
});
test("should match the snapshot", () => {
  const layoutBtnsSnap = renderer.create(
    <LayoutProvider>
      <LayoutChangeBtns />
    </LayoutProvider>
  );
  expect(layoutBtnsSnap).toMatchSnapshot();
});
