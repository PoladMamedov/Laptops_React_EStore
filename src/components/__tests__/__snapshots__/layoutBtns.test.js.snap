// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div
  className="btn-group"
  data-testid="layout-btns"
>
  <input
    checked={true}
    className="btn-check"
    readOnly={true}
    type="checkbox"
  />
  <label
    className="btn btn-outline-secondary"
    onClick={[Function]}
  >
    Cards
  </label>
  <input
    checked={false}
    className="btn-check"
    readOnly={true}
    type="checkbox"
  />
  <label
    className="btn btn-outline-secondary"
    onClick={[Function]}
  >
    List
  </label>
</div>
`;
